### PR TITLE
Allow countermeasures to decoy missiles once only (alternate solution)

### PR DESCRIPTION
**Pull request by niffiwan**
Adds ai_profiles flag
    $countermeasures decoy a missile once:
Also (re)adds logging to make it easier to figure out what the
countermeasures are doing. Try adding "+countermeasures" to your
debug_filter.cfg file.

This is an alternative solution to #415, it ditches the old two CM ignore/chase variables in favour of a size 1 vector for retail compatibility.

### CODE REVIEW

File: code/weapon/weapons.cpp [hunk 1]
```diff
@@ -4101,7 +4106,27 @@ void find_homing_object_cmeasures_1(object *weapon_objp)
 
 				if (dist < cm_wip->cm_effective_rad)
 				{
-					float	chance;
+					float chance;
+
+					if (wp->cmeasure_ignore_list == nullptr) {
+						wp->cmeasure_ignore_list = new SCP_vector<int>;
+						wp->cmeasure_ignore_list->reserve(cm_wip->shots * 2); // guesstimate the number of CMs this weapon will see
```
Comment (MageKing17) [inline code/weapon/weapons.cpp]: We don't need to reserve that many slots if the AI profile flag isn't set.

File: code/weapon/weapons.cpp [hunk 2]
```diff
@@ -4109,17 +4134,27 @@ void find_homing_object_cmeasures_1(object *weapon_objp)
...
-							wp->cmeasure_ignore_objnum = objp->signature;	//	Don't process this countermeasure again.
-						} else  {
-							wp->cmeasure_chase_objnum = objp->signature;	//	Don't process this countermeasure again.
+
+					if (The_mission.ai_profile->flags2 & AIPF2_COUNTERMEASURES_DECOY_ONCE) {
+						wp->cmeasure_ignore_list->push_back(objp->signature);
+					}
+					else {
+						// emulate retail by only "remembering" one seen CM
+						if (wp->cmeasure_ignore_list->size() == 0) {
```
Comment (MageKing17) [inline code/weapon/weapons.cpp]: Wouldn't `wp->cmeasure_ignore_list->empty()` be more concise?

**Comment by MageKing17** [comment]
I don't think this can behave identically to retail because the old behavior tracked two countermeasures: the one it ignored last, *and* the one it chased last. This will only keep track of one of those (e.g. retail, if seeing exactly two countermeasures, ignoring one and chasing the other, will not then "fluctuate" in the future unless it sees a new countermeasure).

**Comment by The-E** [comment]
Is that important? I think the important parts for retail game balance here is the frequency with which countermeasures are evaluated; I do not think that what happens after a missile has been spoofed away from a player or enemy actually matters that much.

**Comment by MageKing17** [comment]
If the only important part is evaluation frequency, than there's no reason for the "$countermeasures decoy a missile once:" flag at all and we should just make the vector default behavior. My point is that of these two PRs, this one cannot actually replicate retail behavior (as-is, anyway).

**Comment by Goober5000** [comment]
I'll quote myself from #415:

> I wonder if that's really a showstopper considering how many other flaws we've found in the countermeasure code. The missile should never have a chance to oscillate if there is only a two-frame limit under retail behavior.
> 
> I value very highly the preservation of retail code when appropriate, but for countermeasures it may not be meaningful.

**Comment by MageKing17** [comment]
For the record, I am actually in favor of doing exactly what I said and just adopting the vector-based tracking for everything, since the balance change is insignificant compared to the global countermeasure timer.

File: code/weapon/weapons.cpp (hunk 1 above)
Comment (niffiwan) [inline code/weapon/weapons.cpp]: How about removing the reserve entirely? I'm starting to think that predicting the number of CMs seen by each missile is practically impossible, and adding conditionals just clutters things up.

File: code/weapon/weapons.cpp (hunk 2 above)
Comment (niffiwan) [inline code/weapon/weapons.cpp]: Yes, I'll fix pending an answer below about removing the AI profiles flag entirely.

**Comment by niffiwan** [comment]
If everyone is happy with the approach, I can remove the AI Profiles flag & leave the vector to remember all the missiles instead? Further away from retail, but probably will not be noticeable per @MageKing17s' comment, and code will be somewhat simpler.

**Comment by The-E** [comment]
Yeah, let's do that.

**Comment by Goober5000** [comment]
I'm fine with that as well.  Yes, countermeasure memory is effectively irrelevant if it only applies the two frames following Alpha 1 launching one.

**Comment by chief1983** [comment]
It looks like the commit from Nov 26th addressed his last proposal, is this waiting on anything else then?

**Comment by MageKing17** [comment]
Not AFAIK, and the code LGTM.

**Pull request merged**